design-documents/frontend/server-side-rendering.md

```diff
@@ -0,0 +1,129 @@
+
+If we leverage the HTTP server to do as much heavy lifting as we can, some of this heavy lifting (non-sensitive pages) can be cached as well, resulting in a faster, better user experience.
+
+### Without SSR
+
+1. A web browser (client) requests a URL for a particular route (e.g., `/`).
+1. The HTTP server runs the appropriate code for that route.
+   - HTML is partially generated with empty containers that need to be filled and rendered by JavaScript after the initial page load (bad for SEO).
+1. The web browser (client) receives and processes the HTML, but the user may see some empty content that hasn't been filled in yet.
+1. JavaScript is loaded and processed, including the React library.
+1. Ideally, a single GraphQL API request is made to collect the data required to fill in the empty containers with content.
+1. React renders the final HTML into the above-mentioned empty containers by combining the data received with the HTML defined in the React components.
+1. Once content is seen, it becomes immediately interactive.
+
+### With SSR
+
+1. A web browser (client) requests a URL for a particular route (e.g., `/`).
+1. The [JavaScript-capable HTTP server](#requirements) runs the appropriate code for that route.
+   - Ideally, a single GraphQL API request is made from this HTTP server, collecting all the data that's required to render the page.
+   - HTML is fully generated [via React methods](https://reactjs.org/docs/react-dom-server.html) with no empty containers. This HTML can be [streamed](https://reactjs.org/docs/react-dom-server.html#rendertonodestream) and cached.
+1. The web browser (client) receives and processes the HTML.
+   - No additional HTML needs to be rendered into containers. As a result, the user perceives faster performance.
+   - At this point, some content might appear to be interactive (e.g., buttons), but won't actually work yet.
+1. JavaScript is loaded and processed, including the React library.
+1. React [hydrates](https://reactjs.org/docs/react-dom.html#hydrate) any containers whose HTML contents were rendered by [ReactDOMServer](https://reactjs.org/docs/react-dom-server.html) in step 2 (above). Nothing needs to be re-rendered here.
+1. The page is now fully interactive.
+1. Subsequent API requests are made directly to the GraphQL API.
+
+## Requirements
+
+- A JavaScript-capable HTTP server, typically a [Node.js][] server (e.g., [Express][]).
+
+## Implementation
+
+Some solutions provide out-of-the-box SSR, like [Next.js](https://nextjs.org/), [Razzle][] and the light-weight [Preact](https://preactjs.com/guide/v10/server-side-rendering); however, if you want more tight control of your application, you might want to set it up yourself. If you go the latter route, you would still do well to use these projects as inspiration. In fact, [Razzle][] has [a great development setup](https://github.com/jaredpalmer/razzle#how-razzle-works-the-secret-sauce).
```

Comment:
Don't think `Next.js` and `Razzle` are comparable technologies.

`Next.js` solves the whole problem of "how do I fetch my data on the server, render it to the client, then pick that back up?"

`Razzle` just runs your webpack builds for you in a way that makes it easy to work on client and server at the same time locally.

Reply:
The only comparison I made between them is that they both "provide out-of-the-box SSR." Are you claiming that this is not the case? What about [this](https://github.com/jaredpalmer/razzle/blob/25f61bc8ce90553747d2335587af653304c085c5/packages/create-razzle-app/templates/default/src/server.js#L15)?

Razzle [makes the following claim](https://github.com/jaredpalmer/razzle#readme):

> Razzle is a tool that abstracts all complex configuration needed for SSR into a single dependency.

I think that's more than "just locally." Granted, Razzle is not the most popular option, but it has some golden nuggets in its configuration.

As for data fetching, you're right. Razzle doesn't handle that, but that's intentional. I'm not really telling anyone what to use here; only, that it can and should be done.

Reply:
I'm going to add a "data fetching" section to elaborate on some of these details.